modules/host/network.go

```diff
@@ -85,9 +85,18 @@ func (h *Host) initNetworking(address string) (err error) {
 // threadedHandleConn handles an incoming connection to the host, typically an
 // RPC.
 func (h *Host) threadedHandleConn(conn net.Conn) {
-	// We defer the conn.Close so that the conn is still closed if `Add` returns
-	// an error.
-	defer conn.Close()
+	// Close the conn immediately on host.Close.
+	defer func() {
+		// We defer the conn.Close so that the conn is still closed if `Add` returns
+		// an error.
+		defer conn.Close()
+		err := h.tg.Add()
+		if err != nil {
+			return
+		}
+		defer h.tg.Done()
+		<-h.tg.StopChan()
+	}()
```

Comment:
Oops, definitely meant it to be `go` not `defer`. Why remove the `Add` and `Done` from the goroutine?

I didn't know that a `net.Conn` can be safely closed twice. Regardless it's still possible to only close it once.
```go
connCloseChan := make(chan struct{})
defer func() { close(connCloseChan) }()
go func() {
    defer conn.Close()
    err := h.tg.Add()
    if err != nil {
        return
    }
    defer h.tg.Done()
    select {
    case <-h.tg.StopChan():
    case <-connCloseChan:
    }
}
```